### ISSUE TITLE
Please provide a flatpak

### ISSUE THREAD

**Issue by leofontenelle** [damonlynch/rapid-photo-downloader#83]
Thanks for the app, it does deliver what the name promises. I would like to use it in a few computers at home which are running Endless OS. Unfortunately, this Linux distro lacks a package manager: the system is updated via OSTree, and packages are installed through flatpaks, usually via Flathub. Rapid Photo Downloader's installation script does not work out of the box in Endless OS, and the missing dependencies can't be installed without turning Endless OS into some sort of Linux From Scratch. In summary, please provide a flatpak for the app, directly or (preferably) via Flathub. Thanks!

**Comment by damonlynch** [comment]
You are welcome to implement the flatpak support yourself. If it works well and is sufficiently robust, it can be incorporated into the project.

I cannot do this because I injured my hands while working on the source code for the program earlier this year. I am dictating this response with a voice recognition program.

**Comment by yodatak** [comment]
Are you open for a Pr?
I can try to make it

**Comment by damonlynch** [comment]
@yodatak yes.

**Comment by yodatak** [comment]
If some people want to help i do some try here 
https://github.com/damonlynch/rapid-photo-downloader/pull/111

**Comment by yodatak** [comment]
For now it crash on 

flatpak run net.damonlynch.rapid_photo_downloader                                                                                                                                                                                                                                                                
Traceback (most recent call last):
  File "/app/bin/rapid-photo-downloader", line 5, in <module>
    from raphodo.rapid import main
  File "/app/lib/python3.10/site-packages/raphodo/rapid.py", line 70, in <module>
    gi.require_version("Notify", "0.7")
  File "/app/lib/python3.10/site-packages/gi/__init__.py", line 126, in require_version
    raise ValueError('Namespace %s not available' % namespace)
ValueError: Namespace Notify not available

maybe some depencies missing for ginspection not realy sure i'm not a developer what do you think @damonlynch  ?

**Comment by damonlynch** [comment]
@yodatak All the software requirements are listed in the README.

Keep in mind that for the program to even operate, any confinement barrier needs to be opened so that libgphoto2 can detect and communicate with cameras, udisks2 can detect and communicate with hard drives, and probably a few more like that.

**Comment by marcoskv** [comment]
From my understanding, and please correct me if I'm wrong:
- building the Flatpak should be straightforward
- issues occur at runtime when accessing the SD card, camera, or other device due to the app'sconfinement
- there are ways to resolve these restrictions, but they may require rewriting parts of the app

It is possible to grant certain permissions to Flatpak applications (see example below).

<img width="734" height="1879" alt="Image" src="https://github.com/user-attachments/assets/d7815da4-cb6f-42d8-ab1c-b30f2a0d8beb" />

I think it would be worth trying to build a Flatpak package, allowing some of these permissions, and identifying the minimum set of permissions required for the app to function properly.

I understand that restricting the app is beneficial from a security perspective, and allowing permissions should be limited as much as possible. However, I currently use Rapid Photo Downloader in two computers (one with Fedora Workstation and the other with Fedora Silverblue).  and I had to use toolbox to install it. Previously, I used the python script for the installation. I do not believe a Flatpak app with granted permissions is very different from the other installation methods mentioned.

However, even with those permissions granted, the app may still not work. Has anyone already tried this? If not, I'm willing to build the Flatpak, test it with my gears (a camera and a SD reader), and try to identify the minimum set of permissions needed. 

Thanks for any suggestion!

**Comment by damonlynch** [comment]
Thank you for your interest in a flatpak version. You are welcome to try it out. A note of caution: it is a lot harder than it looks due to difficult confinement issues. Furthermore, the number of dependencies the program requires is surprising, requiring a lot of work in and of itself.

To get an idea of the complexity involved with respect to confinement, see this technical discussion from six years ago [regarding a possible snap](https://forum.snapcraft.io/t/rapid-photo-downloader-confinement-is-classic-the-only-option/16091/3). 

To get another sense of the complexity, [view the code here](https://github.com/damonlynch/rapid-photo-downloader/blob/main/raphodo/storage/storage.py) that deals with responding to devices being mounted under GIO/GVFS and non-GIO/GVFS systems, including the unmount operations (which are directly related to the previous paragraph's discussion).

The bottom line is that the flatpak confinement must allow Rapid Photo Downloader to use `libgphoto2` to gain exclusive control over the camera. This means Gnome and KDE must be forced to relinquish their control of it (that's how libgphoto2 works). If not, that's a bad outcome. I guarantee that will result in users saying "the program doesn't work", and they will rate it one star out of five, because that's what they've done repeatedly in the past. Similarly, responding to new devices like hard drives and memory cards being plugged and unplugged is also vital, or else users will say Rapid Photo Downloader does not work.

**Comment by marcoskv** [comment]
I managed to build a working Flatpak of v0.9.37. On Fedora 44 Workstation (GNOME 50, Wayland), I tested the following:
- SD card is detected and mounted automatically.
- Pictures are downloaded from SD card
- Connected my Canon 700D with USB cable. Camera is detected automatically, but to be mounted I had to eject it from the system file manager and click "Try Again" in RPD
- Pictures are downloaded from camera
- Connected GoPro Hero 7 with USB cable. Same behavior as the camera: the device is detected, but to be mounted I had to eject it from the system file manager and then click "Try Again" in RPD. GoPro is mounted via MTP
- Connected my Sony Xperia 1 IV with USB cable. Same behavior as GoPro and Camera, but I did not manage to mount it (it says MTP+ADB as connection). Need to run some more tests for this

Other things I should test?

Ah, some more details:
- "Show in File Manager" does not work. This would require changes in the code (need to use `org.freedesktop.portal.OpenURI`, see [here](https://flatpak.github.io/xdg-desktop-portal/docs/doc-org.freedesktop.portal.OpenURI.html)).
- I skipped iOS support for now, since I do not have any Apple device at home. I can add it for the next build, but then we need someone to test it
- Not sure if this is useful or not, but in the logs I see some lines like:
```
QSocketNotifier: Can only be used with threads started with QThread
...
qt.qpa.wayland: Wayland does not support QWindow::requestActivate()
```


How do you suggest proceeding? Should I make a pull request as is? Or should we try to fix the detected problems first? Thanks

**Comment by damonlynch** [comment]
Thank you very much for the work so far! We can continue the discussion here, at this point.

1. Is the SD card is detected and mounted automatically under KDE? KDE does not use GVFS/GIO. Rapid Photo Downloader uses UDisks2 when GVFS/GIO is not running.
2. Does Flatpak provide any mechanism to handle gphoto2? If not, does GVFS/GIO provide a D-Bus interface? 
3. From the looks of things, `org.freedesktop.portal.OpenURI` does some of what Show in File Manager does, but not all. Namely, it does not select the files that are opened. It looks like it can only open the containing directory. Is that your understanding?
4. iOS support is important. I have access to an old iPhone to test it with.
5. You can ignore the QSocketNotifier and qt.qpa.wayland, as they are related to PyQt5. I am in the middle of porting to PyQt6.

Please accept my apologies in advance if I cannot always respond in a timely manner. I have serious health problems that can make it difficult for me to work on this project.
